### PR TITLE
Issue #929: CI Platform abstraction.

### PR DESCRIPTION
**Pull request by drupol**
Preliminary work.

This PR:

- [ ] Enable Github Actions CI
- [ ] Covered by tests

Fixes https://github.com/infection/infection/issues/929

### CODE REVIEW

**Comment by maks-rafalko** [comment]
> now I need to know how you plan to get the CI autodetection working so I can work on it.

Could you please elaborate on that? Not sure what do you mean. 

In order to understand what CI platform to use, we need some kind of factory that would return any of the `Platform` implementation. 

```php
# factory pseudocode

if (getenv('TRAVIS') === 'true') {
    return new TravisPlatform(...);
}

if (getenv('GITHUB_ACTION') === 'true') {
    return new GithubPlatform(...);
}

if (getenv('CIRCLECI') === 'true') {
    return new GithubPlatform(...);
}

return null; // we are on the unsupported platform
```

* Travis https://docs.travis-ci.com/user/environment-variables/#default-environment-variables
* Github https://help.github.com/en/actions/automating-your-workflow-with-github-actions/using-environment-variables#default-environment-variables
* CircleCI https://circleci.com/docs/2.0/env-vars/#built-in-environment-variables

**Comment by BackEndTea** [comment]
What about a 'chain' implementation that takes an array of possible CI classes, and returns the first one that is valid.

File: src/Logger/CI/Github.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    public function getRepositorySlug(): string
+    {
+        return getenv('GITHUB_REPOSITORY');
+    }
+
+    public function getBranchName(): string
+    {
+        return getenv('GITHUB_REF');
```
Comment (hpatoio) [inline src/Logger/CI/Github.php]: Quoting https://help.github.com/en/actions/automating-your-workflow-with-github-actions/using-environment-variables

> The branch or tag ref that triggered the workflow. For example, refs/heads/feature-branch-1. If neither a branch or tag is available for the event type, the variable will not exist.

So this should be something like

```suggestion
        return substr(getenv('GITHUB_REF'), 11, strlen(getenv('GITHUB_REF')));  
```

**Comment by localheinz** [comment]
For reference, also see https://github.com/codeclimate/php-test-reporter/blob/master/src/TestReporter/Entity/CiInfo.php.

**Comment by Lctrs** [comment]
There is also https://github.com/OndraM/ci-detector. It has support for branch name, but it will require a PR for detecting if we're in a PR and to get the repository slug.

**Comment by Lctrs** [comment]
PR to get repository name is here : https://github.com/OndraM/ci-detector/pull/42

**Comment by Lctrs** [comment]
Version 3.2.0 of OndraM/ci-detector is out with everything needed to implement this :)

**Comment by localheinz** [comment]
Also see https://github.com/OndraM/ci-detector/pull/55.

**Comment by maks-rafalko** [comment]
Closed in favor of #1135

Thank you!

**Pull request closed without merging**